OPRUN-4133: OLMv1 single/own namespace install mode support (#1849)
openshift/enhancements: enhancements/olm/single-ownnamespace-enhancement-proposal.md

#### Story 2: Security-Conscious Deployments
As a security-conscious administrator, I want to install operators that explicitly limit their scope to specific namespaces so that I can maintain strict RBAC boundaries and reduce the attack surface of operator deployments.

#### Story 2: Operator Author Requirements
As an operator developer, I have operators designed for namespace isolation that only support Single or OwnNamespace modes, and I want my customers to be able to deploy these operators in OpenShift with OLM v1 so that I can maintain the intended isolation boundaries while benefiting from OLM v1's simplified management.

### Goals

- Enable installation of registry+v1 bundles that only support SingleNamespace or OwnNamespace install modes
- Provide a configuration mechanism through the ClusterExtension API to specify watch namespaces
- Maintain backward compatibility with existing OLM v0 operator content
- Generate appropriate RBAC resources scoped to the target namespaces

> - Support seamless migration path for customers moving from OLM v0 to OLM v1

Author: I added the point about the support for helm-based bundles, but is conversation about SubscriptionConfig on registry+v1 bundles appropriate here?

### Non-Goals

- Re-introducing multi-tenancy features or supporting multiple installations of the same operator
- Supporting MultiNamespace install mode (watching multiple namespaces)
- Modifying the fundamental OLM v1 architecture or adding complex multi-tenancy logic
- Supporting install mode switching after initial installation as a first-class feature
- Enabling this feature for bundle formats other than registry+v1

## Proposal

Update the OLMv1 operator-controller to:
1. Validate bundle compatibility with the requested install mode during resolution
2. Parse the `watchNamespace` configuration from ClusterExtension.spec.config.inline

> 3. Determine the install mode based on the relationship between install namespace and watch namespace

Reviewer: 4. Validate the watchNamespace configuration against the install modes supported by the bundle

4. Generate appropriate RBAC resources (Roles/RoleBindings vs ClusterRoles/ClusterRoleBindings) based on the determined install mode
5. Configure operator deployments with correct environment variables and RBAC for the target watch namespace

Add support for namespace-scoped operation by ensuring:
1. Bundle validation confirms the operator supports the requested install mode
2. RBAC resources are properly scoped to the watch namespace for Single/OwnNamespace modes
3. Operator deployment environment variables are configured for the specified watch namespace

> 4. Installation fails gracefully with clear error messages when bundles don't support the requested mode

Reviewer: 5. Use reasonable default install mode when possible:
 - AllNamespaces by default (when available)
 - OwnNamespace by default (when no AllNamespaces and available)

- Must be explicitly enabled via `--feature-gates=SingleOwnNamespaceInstallSupport=true`

### Topology Considerations

#### Hypershift

OLMv1 does not yet support Hypershift. Although no aspects of this feature's implementation stands out as at odds with the topology of Hypershift, it should be reviewed when OLMv1 is ready to be supported in Hypershift clusters.

#### Standalone Clusters
- **Full Compatibility**: Complete support for Single/OwnNamespace modes
- **Standard RBAC**: Normal Role/RoleBinding generation for namespace-scoped permissions

#### Single-node Deployments
- **Resource Efficiency**: Namespace-scoped operators can reduce resource overhead on constrained single-node deployments
- **Isolation Benefits**: Provides additional isolation even in single-node environments

#### MicroShift

OLMv1 does not yet support Microshift. Although no aspects of this feature's implementation stands out as odds with the topology of Microshift, it should be reviewed when OLMv1 is ready to be supported in Microshift clusters. 

### Implementation Details/Notes/Constraints

The ClusterExtension CRD used in the registry+v1 bundle installation process is enhanced to:

> - Parse watchNamespace configuration from the inline JSON configuration

Reviewer: - Validate watchNamespace against the bundle's supported install modes (maybe we could add a table here to show how validation works for each install mode support configuration)

- Map install modes based on namespace relationships (install vs watch namespace)

> - Generate appropriate RBAC resources scoped to the target namespace

Author: Added a table, PTAL.

# OLM v1 Single and OwnNamespace Install Mode Support

| Title                  | OLM v1 Single and OwnNamespace Install Mode Support                                    |
|------------------------|-----------------------------------------------------------------------------------------|
| Authors                | [@anbhatta]                                                                            |
| Reviewers              | [@joelanford], [@perdasilva], [@thetechnick]            |
| Approvers              |                                                                                        |
| API Approvers          |                                                                                        |
| Creation Date          | 2025-01-20                                                                             |
| Last Updated           | 2025-01-20                                                                             |

> | Tracking Link          | [OPRUN-4133](https://issues.redhat.com/browse/OPRUN-4133)                             |

Reviewer: To fix some linting errors

```suggestion
---
title: single-ownnamespace-enhancement-proposal
authors:
  - anbhatta
reviewers:
  - perdasilva
  - joelanford
approvers:
  - joelanford
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - JoelSpeed
  - everettraven 
creation-date: 2025-09-19
last-updated: 2025-10-21
tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
  - https://issues.redhat.com/browse/OPRUN-4133
---

# OLM v1 Single and OwnNamespace Install Mode Support
```

- Provide a configuration mechanism through the ClusterExtension API to specify watch namespaces
- Maintain backward compatibility with existing OLM v0 operator content
- Generate appropriate RBAC resources scoped to the target namespaces
- Support seamless migration path for customers moving from OLM v0 to OLM v1

#### Future Goals

- Support similar watch namespace configuration of Helm-based bundles
- Expand the configuration surface for registry+v1 bundles to support OLMv0 SubscriptionConfig-type behavior, enabling broader operator configuration compatibility during the migration from OLMv0 to OLMv1 

### Non-Goals

- Re-introducing multi-tenancy features or supporting multiple installations of the same operator
- Supporting MultiNamespace install mode (watching multiple namespaces)
- Modifying the fundamental OLM v1 architecture or adding complex multi-tenancy logic
- Supporting install mode switching after initial installation as a first-class feature
- Enabling this feature for bundle formats other than registry+v1
- Changes to OLM cache infrastructure: OLM will continue to watch bundle contents cluster-scope

> - First-class support for switching install modes: Limited support for mode transitions

Reviewer: Same - let's just remove to avoid confusion and call it out with more words later

### Goals

- Enable installation of registry+v1 bundles that only support SingleNamespace or OwnNamespace install modes
- Provide a configuration mechanism through the ClusterExtension API to pass opaque watchNamespace configuration for bundle rendering
- Maintain backward compatibility with existing registry+v1 bundle content from OLM v0 catalogs
- Generate appropriate RBAC resources scoped to the target namespaces for registry+v1 bundles only
- Support seamless migration path for customers moving from OLM v0 to OLM v1

#### Future Goals

- Expand the configuration surface for registry+v1 bundles to support OLMv0 SubscriptionConfig-type behavior, enabling broader operator configuration compatibility during the migration from OLMv0 to OLMv1 

### Non-Goals

- Re-introducing multi-tenancy features or supporting multiple installations of the same operator
- Supporting MultiNamespace install mode (watching multiple namespaces)
- Modifying the fundamental OLM v1 architecture or adding complex multi-tenancy logic

> - Supporting install mode switching after initial installation as a first-class feature

Reviewer: Maybe we could remove this from dot point from here and point out (with more words) that changing the watchNamespace may work differently for different bundles. I.e. ClusterExtension allows you to change it, it computes the correct resource set, it pivots towards that resource set, but whether that ends up in a working application if application dependent...?

*   **Namespace Resource Quota Impact:** Operators and their RBAC resources count against namespace resource quotas, potentially causing quota exhaustion that doesn't occur with cluster-scoped installations.

### Possible Failure Modes

**Configuration Issues:**
- Invalid watchNamespace specification (DNS1123 validation failures)
- Target namespace doesn't exist or isn't accessible
- ServiceAccount lacks sufficient permissions for namespace access
- Bundle configuration does not include `watchNamespace`

**Runtime Issues:**
- Operator deployed in install namespace but cannot access watch namespace
- RBAC resources incorrectly scoped for actual operator requirements
- Network policies preventing cross-namespace access when needed

### OCP Teams Likely to be Called Upon in Case of Escalation

1. OLM Team (primary)
2. OpenShift API Server Team 
3. Networking Team (cross-namespace connectivity)
4. Authentication & Authorization Team (ServiceAccount/RBAC)
5. Layered Product Team

## Support Procedures

If there are problems with namespace-scoped operator installations:

> 1. **Verify Feature Gate**: Ensure `SingleOwnNamespaceInstallSupport` is enabled

Reviewer: Maybe call out how exactly? e.g. check the the command flags in the operator-controller Deployment in the penshift-operator-controller namespace for the SingleOwnNamespaceInstallSupport=true in the --feature-flags options (double-check the actual option name though)

| ✓             | ✓               | -            | optional. If set, must NOT be install namespace (default: unset) | 
| ✓             | ✓               | ✓            | optional (default: unset)                                        | 

Add support for namespace-scoped operation by ensuring:
1. Bundle validation confirms the operator supports the requested install mode
2. RBAC resources are properly scoped to the watch namespace for Single/OwnNamespace modes
3. Operator deployment environment variables are configured for the specified watch namespace
4. Installation fails gracefully with clear error messages when bundles don't support the requested mode
5. Usage of reasonable default install mode when possible: 
  - AllNamespaces by default (when available)
  - OwnNamespace by default (when no AllNamespaces and available)

Ensure parity with OLMv0 behavior by:
1. Generating identical RBAC and deployment configurations for equivalent install modes
2. Maintaining the same namespace isolation boundaries
3. Preserving operator functionality across install mode transitions

### Workflow Description

#### Administrator Workflow

> 1. **Enable Feature Gate**: Administrator enables the `SingleOwnNamespaceInstallSupport` feature gate on the operator-controller deployment

Reviewer: We should either remove this (as it will be enabled in GA), or call out what "while in TP" and use the downstream feature gate name `NewOLMOwnSingleNamespace`

Author: That's a good point, removing this altogether so that the documentation in other places can take care of listing if this step is needed or not.

2. **Create ServiceAccount**: Administrator creates a ServiceAccount with appropriate permissions in the target namespace
3. **Configure ClusterExtension**: Administrator creates a ClusterExtension resource specifying:
   - `spec.namespace`: The installation namespace where the operator pod will run
   - `spec.config.inline.watchNamespace`: The namespace the operator should watch for resources

> 4. **Install Mode Detection**: The system automatically determines the install mode according to the citerion table enumerated in the `Implementation Details/Notes/Constraints` section. 

Reviewer: Maybe we could add something around "configuration is validated based on the bundle configuration" (e.g. AllNamespaces-only bundles do not contain any configuration)

> A scenior exists where the user must specify the watch namespace. Example workflow of a bundle that will require the watch namespace to be specified: 

Reviewer: ```suggestion
A scenario exists where the user must specify the watch namespace. Example workflow of a bundle that will require the watch namespace to be specified: 
```

	// +kubebuilder:validation:Required
	ConfigType ClusterExtensionConfigType `json:"configType"`

	// inline contains JSON or YAML values specified directly in the
	// ClusterExtension.
	//
	// inline must be set if configType is 'Inline'.
	//
	// The Type=object validation ensures that inputs are valid key/value
	// configuration objects, rejecting valid JSON that isn't configuration-like
	// (e.g., boolean `true`/`false`, arrays, or strings) and enforcing object structure
	// resembling configuration files with key: value pairs.
	//
	// +kubebuilder:validation:Type=object
	// +optional
	Inline *apiextensionsv1.JSON `json:"inline,omitempty"`
}
```

```
apiVersion: olm.operatorframework.io/v1
kind: ClusterExtension
metadata:
  name: argocd
spec:
  namespace: argocd
  serviceAccount:
    name: argocd-installer
  config:
    inline:
      watchNamespace: argocd-pipelines
  source:
    sourceType: Catalog
    catalog:
      packageName: argocd-operator
      version: 0.6.0
```

Initially, only the `Inline` configType will be available. However, we leave it expandable in case further 
configuration sources (e.g. ConfigMaps, Secrets, etc.) become needed.

#### Feature Gate

> The feature is controlled by the `SingleOwnNamespaceInstallSupport` feature gate:

Author: Replaced this section with the sentence "This feature can be enabled via the feature gate `NewOLMOwnSingleNamespace` when the feature is in tech preview."

| All + Single + Own      | Specified (= install ns) | OwnNamespace          | Namespace relationship determines mode |
| All + Single + Own      | Specified (≠ install ns) | SingleNamespace       | Namespace relationship determines mode |

Default Mode Selection Rules are as follows:
  1. No watchNamespace specified: Use the most permissive mode supported by the bundle (AllNamespaces > OwnNamespace >
  SingleNamespace)
  2. watchNamespace = install namespace: Use OwnNamespace mode if supported, otherwise error
  3. watchNamespace ≠ install namespace: Use SingleNamespace mode if supported, otherwise error
  4. Unsupported mode requested: Installation fails with clear error message indicating supported modes

The relevant generated resources will be:
- Role/RoleBinding resources for namespace-scoped permissions in Single/OwnNamespace modes
- ClusterRole/ClusterRoleBinding resources for cluster-scoped permissions (AllNamespaces mode)
- Deployment resources with correct environment variables for watch namespace configuration
- Service and other supporting resources in the install namespace

> Feature gate `SingleOwnNamespaceInstallSupport` controls:

Author: Did the same thing as the other place. Removed this paragraph altogether since it was repeating information that's already enumerated above, while saying "they'll be behind a feature gate".

    *   Example: Namespace admin removes critical RoleBinding, causing operator to lose access to required resources.

**Resource Utilization:**

*   **RBAC Resource Proliferation:** Each Single/OwnNamespace installation creates namespace-scoped RBAC resources instead of reusing cluster-scoped ones. Multiple operators in different namespaces will create duplicate Role/RoleBinding resources rather than sharing ClusterRole/ClusterRoleBinding resources.
    *   Example: Installing the same operator in 10 different namespaces creates 10 sets of Role/RoleBinding resources instead of 1 set of ClusterRole/ClusterRoleBinding resources.
*   **Namespace Resource Quota Impact:** Operators and their RBAC resources count against namespace resource quotas, potentially causing quota exhaustion that doesn't occur with cluster-scoped installations.

### Possible Failure Modes

**Configuration Issues:**
- Invalid watchNamespace specification (DNS1123 validation failures)
- Target namespace doesn't exist or isn't accessible
- ServiceAccount lacks sufficient permissions for namespace access

> - Bundle doesn't support requested install mode

Reviewer: ```suggestion
- Bundle configuration does not include `watchNamespace`
- Given value for `watchNamespace` is not supported by the bundle
```

# OLM v1 Single and OwnNamespace Install Mode Support

title: single-ownnamespace-enhancement-proposal
authors:
  - anbhatta
reviewers:
  - perdasilva
  - joelanford
approvers:
  - joelanford
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"

>   - JoelSpeed

Reviewer 2: For posterity:
```suggestion
```

// +union
type ClusterExtensionConfig struct {
	// configType is a required reference to the type of configuration source.
	//
	// Allowed values are "Inline"
	//
	// When this field is set to "Inline", the cluster extension configuration is defined inline within the
	// ClusterExtension resource.
	//
	// +unionDiscriminator
	// +kubebuilder:validation:Enum:="Inline"
	// +kubebuilder:validation:Required
	ConfigType ClusterExtensionConfigType `json:"configType"`

	// inline contains JSON or YAML values specified directly in the
	// ClusterExtension.
	//
	// inline must be set if configType is 'Inline'.
	// inline accepts arbitrary JSON/YAML objects.
	// inline is validation at runtime against the schema provided by the bundle if a schema is provided.
	//
	// +kubebuilder:validation:Type=object
	// +optional
	Inline *apiextensionsv1.JSON `json:"inline,omitempty"`
}
```

```
apiVersion: olm.operatorframework.io/v1
kind: ClusterExtension
metadata:
  name: argocd
spec:
  namespace: argocd
  serviceAccount:
    name: argocd-installer
  config:
    inline:

>       watchNamespace: argocd-pipelines

Reviewer 2: Another thought to layer on top here, this type of configuration also allows you to make this legacy installation mode option something that is considered in bundle resolution.

Instead of selecting a bundle to install and then identifying that it doesn't support this installation mode you can include the installation mode in your resolution criteria, returning a resolution error if no bundles match those criteria.

>       watchNamespace: argocd-pipelines

Reviewer 3: Lots to unpack here:
1. It is a core design decision of OLMv1 to ensure that watchNamespace is never a first class field in our API. See https://operator-framework.github.io/operator-controller/project/olmv1_design_decisions/#watched-namespaces-cannot-be-configured-in-a-first-class-api. Putting it in a structured field in the API would make it a first class field. Putting any reference to `watchNamespace` or `installMode` in our CRD is a non-starter.
2. The core tenant of this API design is that bundles will define their contents and configuration schemas. The ClusterExtension API _does not_ define the configuration schema. This is important for several reasons.
   1. If we put legacy configuration in one section and another bundle format's config in another section, it means that users would have to have a priori awareness of the bundle format, even though that is an implementation detail.
   1. An open-ended configuration schema allows for bundle format changes without requiring ClusterExtension configuration changes. This is an extremely important detail in the larger migration story away from the registry+v1 bundle format. We want to be able to tell extension authors that want to move to Helm that they can use the JSON schema of their registry+v1 bundle as the values schema of their Helm chart, which would mean they could transparently change bundle formats without extension consumers needing to change their ClusterExtension.
   1. We know from the experience of OLMv0 that restrictive and opinionated configuration schemas defined by the OLM team ultimately cause problems when customers ask extension authors to deliver a new feature related to how the operator itself is configured. We've seen this happen multiple times with the `Subscription.spec.config` and the registry+v1 format. Example include:
      - needing to add [nodeAffinity](https://github.com/operator-framework/api/pull/250) to the subscription.spec.config API in order for customers with multi-arch clusters to properly install their operators.
      - needing to add [annotations](https://github.com/operator-framework/api/pull/312) to the subscription.spec.config API in order for users to use OLM in clusters that have policies that require certain annotations on objects.
      - needing to add NetworkPolicy as a allowed kind in the registry+v1 bundle format. (this is not directly related to runtime configuration, but fits in the overall theme of OLMv1 needing to be less opinionated when it comes to the shape of bundles and their configuration)

@everettraven, I think the main flaw of the argument for admission-time validation is that OLM and the user know what the format of the resolved bundle will be prior to resolution occurring, which happens during reconciliation of the ClusterExtension. Even if we could build an admission webhook that performs resolution and schema validation at admission time, that wouldn't remove the need for runtime validation because:
1. The admission-time resolution might be different than the reconcile-time resolution, which means the admission time validation may be invalid by the time we reconcile and _actually_ resolve.
2. The ClusterExtension can represent a range of versions that may not even exist yet. While it may be technically possible in the future to assess validity of the configuration based on the schemas of all of the bundles that match the filter criteria, it is impossible to assess configuration validity for yet unpublished bundles that will fall into the configured range in the future.

And to reiterate, we did consider and reject the idea of a discriminated union for different bundle formats because of the major friction point that would add for users to move off of our legacy bundle format and onto Helm, which I think I've already seen conceded as a format that would require open-ended configuration.

>       watchNamespace: argocd-pipelines

Reviewer: > Enhancements should be created and reviewed before you've already done the work.

That's true, and I've raised this with Joe and will raise it with the team. It's certainly my first go around with this process, and I think can and will do better in the coming iterations. We'll certainly adapt our internal processes to avoid this situation again.

> I've pointed out more than just the runtime validation being problematic here. If you end up with a multi-format catalog and someone defines a configuration that is compatible with one format but not the other, OLM should not attempt to install an incompatible bundle format.

I agree. It's just not where we want to go. We don't want to have bundle format concerns floating up to the API.

> I do consider this blocking because the decision made here has a direct impact on the quality of the product. It may be easier to come to a consensus in a synchronous meeting.

I agree. But, that's also the reason why we didn't want to have bundle concerns in the API...:shrug: I'm happy to have a synchronous meeting at any time. Is next Tuesday at the API Review office hours still the best time for you?

> If you can get to a world where OLM is totally agnostic to the bundle side of things, which seemingly would mean a new singular bundle format, I can get behind the opaque configuration.

I don't think it necessarily means a new singular bundle format. At least not exclusively. It could also mean that the supported formats need to conform to the same interface, and we're kind of there given that we'll:
 - more than likely support Helm content (maybe not the format), and 
 - we want a new reg+v2 format, and
 - we need to support reg+v1, and
 - the idea is to allow authors to define their own configuration surface, and
 - all of these formats support the opaque configuration/runtime validation model

While we could add legacy configuration with ingress-time validation because registry+v1 allows for that, and we did consider that, we don't want to because it forces the user to have to think about the bundle format - and we believe that has a direct impact on the quality of the product.

In any case, this isn't really a one-way door. We can always add a legacy configuration format if really becomes necessary. Taking it away however, is a lot harder. The opaque option will always be a thing anyway...

I'm gonna jump off this thread. I don't think this back and forth is solving anything. Let's go the synchronous route. From our perspective, the earlier the better. But, if the best time for you is next Tuesday at the office hours, let's just do that...

>       watchNamespace: argocd-pipelines

Reviewer 2: I _do_ think that pushing the responsibility of configuration, templating, etc. to the bundle is a reasonable direction forward, but I think you need an answer for what that actually looks like before designing an API around it and how OLM is going to behave in that new world.

To me, it doesn't seem reasonable to have a singular opaque configuration field in a world where you support multiple distinct bundle formats for packaging.

In a world where you support a single bundle format where the configuration process is entirely offloaded to the bundle, I think it is reasonable.

>       watchNamespace: argocd-pipelines

Reviewer 2: >reject the idea of a discriminated union for different bundle formats because of the major friction point that would add for users to move off of our legacy bundle format and onto Helm

I'm curious, what is the major friction point in switching the configuration in the ClusterExtension API?

>       watchNamespace: argocd-pipelines

Reviewer 2: Had a chat with both @joelanford and @perdasilva to hash this out synchronously.

With a better understanding of the direction you folks are heading and having addressed my concerns about certain scenarios - if you folks feel as though this opaque configuration is what provides the best experience for your users and are ready to support it along with any challenges that come along in the future I won't stop it.

Please make sure that you update the EP as discussed to include:
- An explicit statement about the future you are trying to achieve and how this is the best path to achieving that
- How to ensure customers encountering issues here are getting the right support (i.e who is responsible for what failure modes)
- How users can prevent and/or remediate bad states caused by either:
    - a misconfiguration
    - a scenario where an end user had a valid configuration and an automatic upgrade happens where the configuration is now invalid and blocked
    - a scenario where a user had a valid configuration and an automatic upgrade happens where the  previous configuration is no longer respected (i.e their workflows are now broken because their intended configuration is no longer in place)

## Graduation Criteria

### Dev Preview -> Tech Preview
- [ ] Feature-gated implementation behind `NewOLMOwnSingleNamespace`
- [ ] Basic functionality for Single and OwnNamespace modes
- [ ] Unit and integration test coverage
- [ ] Documentation for configuration and usage

### Tech Preview -> GA
- [ ] 1 OCP release of alpha feedback
- [ ] Production deployment validation
- [ ] Complete documentation including best practices
- [ ] Established support and maintenance processes

### Removing a deprecated feature

NA

## Upgrade / Downgrade Strategy

### Upgrade Strategy
- **Feature Gate Dependency**: Feature must be enabled via feature gate before configuration can be used
- **Backward Compatibility**: Existing AllNamespaces installations continue to work unchanged
- **Configuration Migration**: No automatic migration; users must explicitly install using OLMv1 ClusterExtension and configure `watchNamespace`

### Downgrade Strategy

> - **Feature Gate Disable**: Disabling the feature gate prevents new Single/OwnNamespace installations

Reviewer 2: Customers do not have direct control of OpenShift features at a granular level. They can choose to flip to an unsupported mode that enables the ones in that level of maturity.

The goal of this section is to understand how, if for some reason, a customer would need to go about downgrading to a version where your feature is no longer enabled by default.

In my experience, for features that involve new opt-in API fields this ends up being something like:
- unset new fields
- downgrade

In your case, I'd include some considerations for how OLM will handle a ClusterExtension using a field it doesn't know about any more on a downgrade and what things a customer doing a downgrade may need to consider before going through with it.

- **Operational Impact:**
  - Permission debugging requires understanding of install mode impact on RBAC scope
  - Security auditing must consider namespace-level vs cluster-level permission grants
  - Upgrade scenarios may change RBAC scope if install mode changes

### Impact on Existing SLIs

With the removal of the install mode concept in olmv1, operator packages that want to continue to use this stop-gap feature are expected to surface configuration documentation, and call out if the `watchNamespace` parameter is a part of it, along with usage example etc. Until a broad percentage of operator pacakges, especially those that don't support `AllNamespaces` mode, take action to make such documentation avaialble, a spike in bad installation is expected.

**Installation Success Rate:** 

*   **RBAC Validation Complexity:** Namespace-scoped installations require more complex RBAC validation to ensure the ServiceAccount has appropriate permissions for the target namespace. RBAC misconfigurations that work in AllNamespaces mode may fail in Single/OwnNamespace modes.

>     *   Example: ServiceAccount has cluster-wide read permissions but lacks namespace-specific write permissions, causing installation to fail.

Reviewer 2: Is this the OLM service account or operand service account you are referring to?

If it is the OLM service account this sounds true regardless of namespace-scoped vs cluster-scoped installations because any extension in this context inherently requires stamping out some kind of namespaced resources.

*   **Bundle Compatibility Validation:** Additional validation layer to confirm bundles support the requested install mode. Bundles that only support AllNamespaces will fail when Single/OwnNamespace is requested.
    *   Example: Attempting to install a bundle with `watchNamespace: "test"` when the bundle CSV only declares support for AllNamespaces install mode.

**Installation Time:**

*   **Extended Validation Phase:** Additional validation steps for namespace existence, accessibility, and RBAC permissions add latency to the installation process. Each namespace-scoped installation must validate the target namespace and ServiceAccount permissions.

> *   **RBAC Generation Complexity:** Converting cluster-scoped RBAC to namespace-scoped Role/RoleBinding resources requires additional processing time. Complex operators with extensive permission requirements will see increased installation duration.

Reviewer 2: Where are you converting cluster scoped RBAC to namespace scoped RBAC in this process? 

The only place I saw mention of a conversion was the CSV `permissions` being translated to cluster scoped RBAC resources when installing as `AllNamespaces` mode - which is the status quo.

> *   **Cross-Namespace Connectivity Validation:** Single namespace mode requires validation that the operator in the install namespace can access resources in the watch namespace, adding network connectivity checks.

Reviewer 2: This isn't something that OLM itself is going to check right?

I'd expect OLM to check whether or not it was given permissions to stamp out the roles/rolebindings in the   watch namespace but not do actual checks for whether or not the operator it is installing is able to access the things it cares about.

The enhancement introduces a configuration mechanism through the ClusterExtension API that allows users to specify a `watchNamespace` parameter as opaque bundle configuration, enabling registry+v1 bundles to be rendered in either SingleNamespace mode (watching a namespace different from the install namespace) or OwnNamespace mode (watching the same namespace where the operator is installed).

## OLM v1 Design Principles and Scope

**This enhancement does NOT change OLM v1's core design principles.** OLM v1 maintains its fundamental stances:

- **No multi-tenancy support**: OLM v1 does not and will not support multi-tenancy features. As stated in the [upstream design decisions](https://operator-framework.github.io/operator-controller/project/olmv1_design_decisions/), "Kubernetes is not multi-tenant with respect to management of APIs (because APIs are global)."
- **No first-class watchNamespace configuration**: OLM v1 does not provide first-class API support for configuring which namespaces a controller watches.
- **Single ownership model**: Each extension is owned by exactly one ClusterExtension to avoid shared ownership complexity.

> 

Reviewer: ```suggestion
- **Format agnostic**: While OLMv1 aims to support a few different content types (e.g. registry+v1, Helm charts, a new olm bundle format), the API is format agnostic and doesn't provide any format specific surfaces.
```

> This enhancement provides **limited compatibility support for existing registry+v1 bundle content only**. The `watchNamespace` configuration is treated as opaque bundle-provided configuration used solely for rendering registry+v1 bundles into appropriate manifests. This is not a multi-tenancy feature but rather a bundle format compatibility layer to support migration from OLM v0 to OLM v1.

Reviewer: ```suggestion
It is important to highlight that the introduction of Single/OwnNamespace support is not intended as a pathway to multi-tenancy. Under OLMv1, an Operator can only be installed once. We also distance ourselves from notions of install modes in v1. They are called out in this enhancement merely for didactic purposes. Instead, in OLMv1, we see configuration as something to be defined by the content author. As such, we treat the support for Single/OwnNamespace modes that some bundles have as configuration concerns outside the purview of OLM. E.g. there is a bundle, it provides a configuration `watchNamespace` that may be required or optional, and may be restricted to, or prevented from, being the install namespace, or can take any valid namespace name.
```

Future OLM v1 bundle formats will not include these legacy install mode concepts, as OLM v1 will not have first-class features or opinions on RBAC management in next-generation bundle formats.

## Motivation

> The registry+v1 bundle format (used by OLM v0) includes install mode declarations (AllNamespaces, MultiNamespace, SingleNamespace, OwnNamespace) that affect how bundles are rendered into Kubernetes manifests. OLM v1 deliberately simplified operator installation by focusing on AllNamespaces mode to avoid complexity, but there exists significant existing registry+v1 bundle content in catalogs that only declares support for Single and OwnNamespace install modes.

Reviewer: ```suggestion
The registry+v1 bundle format (used by OLM v0) includes install mode declarations (AllNamespaces, MultiNamespace, SingleNamespace, OwnNamespace) that affect how bundles are rendered into Kubernetes manifests. OLM v1 deliberately simplified operator installation by focusing on AllNamespaces mode to avoid complexity, but there is significant existing registry+v1 bundle content in catalogs that only declares support for Single and OwnNamespace install modes.
```
